### PR TITLE
feat(plugin): add upgrade functionality to plugin

### PR DESCRIPTION
**Pull request by sinhaashish**

### CODE REVIEW

File: control-plane/plugin/Cargo.toml [hunk 1]
```diff
@@ -36,6 +36,14 @@ serde_yaml = "0.8.26"
 humantime = "2.1.0"
 uuid = { version = "0.8.2", features = ["serde", "v4"] }
 serde_derive = "1.0.140"
+maplit = "1.0.2"
+k8s-openapi = { version = "0.15.0", features = ["v1_20"] }
```
Comment (tiagolobocastro) [inline control-plane/plugin/Cargo.toml]: this is the rest plugin where all non-k8s specific commands go, there should be no k8s/kube usages here

File: control-plane/plugin/src/bin/rest-plugin/main.rs [hunk 2]
```diff
@@ -116,5 +116,10 @@ async fn execute(cli_args: CliArgs) {
                 node::Node::uncordon(id, label, &cli_args.output).await
             }
         },
+        Operations::Upgrade(resource) => match resource {
```
Comment (tiagolobocastro) [inline control-plane/plugin/src/bin/rest-plugin/main.rs]: Right I see, so this upgrade command should not live here in the rest-plugin, only in the kubectl-plugin.

**Comment by sinhaashish** [comment]
bors try

File: k8s/plugin/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+use async_trait::async_trait;
+use std::path::PathBuf;
+
+pub mod resources;
```
Comment (tiagolobocastro) [inline k8s/plugin/src/lib.rs:4]: missing doc comment

File: k8s/plugin/Cargo.toml [hunk 4]
```diff
@@ -27,6 +27,16 @@ anyhow = "1.0.44"
 clap = { version = "3.1.5", features = ["color", "derive"] }
 humantime = "2.1.0"
 shutdown = { path = "../../utils/shutdown" }
+maplit = "1.0.2"
+k8s-openapi = { version = "0.15.0", features = ["v1_20"] }
+actix-rt = "2.7.0"
```
Comment (tiagolobocastro) [inline k8s/plugin/Cargo.toml:32]: do we need this?

File: k8s/plugin/src/main.rs [hunk 5]
```diff
@@ -45,6 +46,12 @@ struct CliArgs {
     /// Kubernetes namespace of mayastor service, defaults to mayastor
     #[clap(global = true, long, short = 'n', default_value = "mayastor")]
     namespace: String,
+
+    /// Endpoint of upgrade operator service, if left empty then it will try to parse endpoint
+    /// from upgrade operator service(K8s service resource), if the tool is unable to parse
+    /// from service then logs will be collected using Kube-apiserver
```
Comment (tiagolobocastro) [inline k8s/plugin/src/main.rs:52]: "logs will be collected using Kube-apiserver"
copy paste mistake?

File: Cargo.lock [hunk 6]
```diff
@@ -87,6 +87,7 @@ version = "2.7.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "7ea16c295198e958ef31930a6ef37d0fb64e9ca3b6116e6b93a8bdae96ee1000"
 dependencies = [
+ "actix-macros",
```
Comment (tiagolobocastro) [inline Cargo.lock:90]: why is this being pulled?

File: control-plane/plugin/src/bin/rest-plugin/main.rs [hunk 7]
```diff
@@ -116,5 +126,13 @@ async fn execute(cli_args: CliArgs) {
...
             }
         },
+        // Operations::Upgrade(resource) => match resource {
+        //     GetUpgradeResources::Install => upgrade::Upgrades::install().await,
+        //     GetUpgradeResources::Apply => upgrade::Upgrades::apply(cli_args.kube_config_path,
+        // cli_args.timeout, cli_args.upgrade_operator_endpoint).await,
+        //     GetUpgradeResources::Get => upgrade::Upgrades::get(cli_args.kube_config_path,
+        // cli_args.timeout, cli_args.upgrade_operator_endpoint).await,
+        //     GetUpgradeResources::Uninstall => upgrade::Upgrades::uninstall().await,
+        // },
```
Comment (tiagolobocastro) [inline control-plane/plugin/src/bin/rest-plugin/main.rs:136]: Please remove commented out code.

File: control-plane/plugin/src/bin/rest-plugin/main.rs [hunk 8]
```diff
@@ -24,6 +24,10 @@ struct CliArgs {
     #[clap(subcommand)]
     operations: Operations,
 
+    /// Path to kubeconfig file.
+    #[clap(parse(from_os_str), global = true, long, short = 'k')]
+    kube_config_path: Option<PathBuf>,
```
Comment (tiagolobocastro) [inline control-plane/plugin/src/bin/rest-plugin/main.rs:29]: remove this and upgrade_operator_endpoint from rest-plugin.

File: k8s/plugin/src/resources/constant.rs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+pub(crate) const UPGRADE_OPERATOR_SERVICE_PORT: i32 = 8080;
+/// Service internal port constant for upgrade operator.
+pub(crate) const UPGRADE_OPERATOR_INTERNAL_PORT: i32 = 8080;
+/// Upgrade inage tag
+//pub(crate) const UPGRADE_IMAGE: &str = "mayadata/mayastor-operator-upgrade:develop";
+
+/// The service port for upgrade operator.
+pub const UPGRADE_OPERATOR_HTTP_PORT: &str = "http";
+/// Upgrade inage tag
+pub(crate) const UPGRADE_IMAGE: &str = "niladrih/mayastor-operator-upgrade:test";
```
Comment (tiagolobocastro) [inline k8s/plugin/src/resources/constant.rs:54]: this image should probably be in utils-lib/src/constants.rs
Also, remember to update this before merging :)

Comment (tiagolobocastro) [reply k8s/plugin/src/resources/constant.rs:54]: Should this be overridable via cmdline?

File: k8s/plugin/src/resources/mod.rs [hunk 10]
```diff
@@ -20,6 +28,42 @@ pub enum Operations {
     /// 'Uncordon' resources.
     #[clap(subcommand)]
     Uncordon(CordonResources),
+    /// 'Upgrade' resources
+    #[clap(subcommand)]
+    Upgrade(GetUpgradeResources),
```
Comment (tiagolobocastro) [inline k8s/plugin/src/resources/mod.rs:33]: hmm I don't this is quite up to the current implementation.
Although I prefer what you did, let's keep with the existing pattern until we change everything, so we'd have all the verbs (Install,Apply,Get,Uninstall) at the root.

File: k8s/plugin/src/resources/constant.rs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        "openebs.io"
+    };
+    ($s:literal) => {
+        concat!(upgrade_group!(), "/", $s)
+    };
+}
+
+/// macros to define labels for upgrade operator.
+#[macro_export]
+macro_rules! upgrade_labels {
```
Comment (tiagolobocastro) [inline k8s/plugin/src/resources/constant.rs:14]: can these macros be made functions instead?

File: k8s/plugin/src/resources/objects.rs [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+                verbs: vec!["get", "list", "watch", "delete"]
+                    .iter()
+                    .map(|s| s.to_string())
+                    .collect(),
+                ..Default::default()
+            },
+            PolicyRule {
+                api_groups: Some(vec!["".to_string()]),
+                resources: Some(vec!["nodes".to_string()]),
+                verbs: vec!["get", "list", "patch"]
```
Comment (tiagolobocastro) [inline k8s/plugin/src/resources/objects.rs:109]: You have a mix of "".to_string() and iter...collect here. 

You can add this to the utils-lib:
```rust
/// Helper to convert from Vec<F> into Vec<T>
pub trait IntoVec<T>: Sized {
    /// Performs the conversion.
    fn into_vec(self) -> Vec<T>;
}

impl<F: Into<T>, T> IntoVec<T> for Vec<F> {
    fn into_vec(self) -> Vec<T> {
        self.into_iter().map(Into::into).collect()
    }
}
```
Then you can simply have vec!["this", "that"].into_vec()

File: k8s/plugin/src/resources/upgrade.rs [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+
+    pub async fn get_service_account(&self) -> ObjectList<ServiceAccount> {
+        let lp = ListParams::default().fields(&format!(
+            "metadata.name={}",
+            UPGRADE_OPERATOR_SERVICE_ACCOUNT
+        ));
+        self.upgrade_service_account
+            .list(&lp)
+            .await
+            .expect("failed to list service accounts")
```
Comment (tiagolobocastro) [inline k8s/plugin/src/resources/upgrade.rs:101]: Please remove "expect" and replace with return of errors.

File: k8s/plugin/src/resources/upgrade.rs [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+            let service_account = objects::upgrade_operator_service_account(ns);
+            let pp = PostParams::default();
+            match self
+                .upgrade_service_account
+                .create(&pp.clone(), &service_account)
+                .await
+            {
+                Ok(_) => {
+                    println!("service account created");
+                    tokio::time::sleep(Duration::from_secs(5)).await;
```
Comment (tiagolobocastro) [inline k8s/plugin/src/resources/upgrade.rs:160]: why are we sleeping?

**Comment by sinhaashish** [comment]
closing this Pr as PR https://github.com/openebs/mayastor-extensions/pull/63 is raised for the same.

**Pull request closed without merging**